ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -0,0 +1,187 @@
+reviewers:
+  - "@qu1queee"
+  - "@HeavyWombat"
+approvers:
+  - "@adambkaplan"
+  - "@sbose78"
+creation-date: 2021-09-10
+last-updated: 2021-09-10
+status: draft
+
+---
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+This ship provides a mechanism to surface errors in the build process
+into BuildRun objects granting more visibility into
+without being Tekton/Implementation specific. To
+achieve this goal, we introduce two new generic results
+`shp-error-reason` and `shp-error-message`, that a TaskRun can emit
+and a BuildRun can pickup.
+
+
+## Motivation
+A strategy's container logs differ for its steps and utilized
+tooling. Therefore, finding an error message and extracting the error
+reason is challenging for machine processing and consumption from
+third parties due to the logs heterogeneous nature and the need to parse
+`stdout`. 
+For a better user experience, we propose that a strategy
+should be able to present the failure reason in a compact way that can
+be represented in a BuildRun CRD, for example, as a TaskRun result. This
+task is more straightforward since a TaskRun is aware of its tooling
+and possible errors, which it can emit similarly to [digest](https://github.com/shipwright-io/build/blob/main/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L65).
+
+Additionally, the proposal enables a user to keep the troubleshooting
+process on the BuildRun level and only descend further into Tekton
+specifics when necessary.
+
+### Goals
+
+- A mechanism to automatically pick up error results from a TaskRun
+  Surface them to the associated BuildRun
+
+### Non-Goals
+
+- The specifics of errors in regards to tooling are not part of this
+  proposal. Neither is their means for reporting them. Errors highly
+  depend on each strategy
+
+## Proposal
+
+We propose to introduce two generic results for errors:
+`shp-error-message` and `shp-error-reason`.  The result
+`shp-error-reason` have a recognizable name that is ideally prefixed
+with the problematic technology, for example, `git` in case of missing
+a secret. The result `shp-error-message` should be descriptive text to
+give the user enough information to troubleshoot the BuildRun.  A task
+run emits the new results as Tekton results.  The BuildRun CRD is
+assumed to have the `status.results` according to `0023` (alternatives
+are discussed further below). 
+Finally, we must extend the BuildRun controller
+logic to propagate the new results into the BuildRun CRD's
+`status.results` property.
+
+BuildRun with new error results: 
+
+```yaml
+status:
+   results:
+       - name: shp-error-reason
+         value: git-ssh-auth
+       - name: shp-error-message
+         value: Authentication failed for ...
```

Comment:
I see the point for having an extra field failure separate from the results field from ship 23. However, I am unsure about the strongly typed approach. Do you propose that we have a new struct with the fields `reason` (string) and `message` (string). Or is `reason` an enum type with specific error reasons.

Reply:
Yes, we will have a new struct with a reason and message. Both are strings because they can be set by the build strategy author. I thought about making reason an enum with us providing a set of options there (generic things like FileNotFound or ParamValueInvalid), this would help a generic client like the `shp` CLI. But, I for me concluded that this does not make sense. The variety of possible errors is way larger than what we will ever be possible to imagine, as such any enum list we would provide can only be incomplete.